### ISSUE TITLE
Issues when using buffers with DDP

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#63916]
https://gist.github.com/rohan-varma/5d599284b632f4abdcf633e720aae9d7 illustrates a possible issue with how DDP synchronizes buffers. The output of the script indicates that module buffers across the 2 ranks are not actually synchronized (i.e. they do not take on rank 0's value as expected) before the forward pass.

It appears that the reason for this is due to the model code which reassigns the buffer tensor:
```
self.buffer = torch.randn(1, 2)
```
Since this reassigns the tensor, the value in `self.modules_buffers` is no longer correct and thus we end up synchronizing values that don't really have any meaning in the model anymore. 

This reassign is motivated from the batchnorm code (but note that the BN code doesn't actually result in a correctness issue since the `num_batches_tracked value` is always the same across ranks anyways). So the current state is that manipulating a buffer in the following way:

`buffer.add_(1)`
would work, but
`buffer = buffer + 1`
would not. This seems like it can a little brittle and prone to errors.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23
